### PR TITLE
8261579: AArch64: Support for weaker memory ordering in Atomic

### PR DESCRIPTION
**Pull request by theRealAph**
At present the Atomic operations in HotSpot only support conservative (the very strongest) and relaxed (the weakest) memory ordering.
We should add at least seq_cst for LSE. This patch also adds a release-only CAS, needed for Shenandoah.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8261579](https://bugs.openjdk.java.net/browse/JDK-8261579): AArch64: Support for weaker memory ordering in Atomic

### Reviewers
 * [Andrew Dinn](https://openjdk.java.net/census#adinn) (@adinn - **Reviewer**) ⚠️ Review applies to 6298e85ae00a461a6938607f583927e855795d25
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4597/head:pull/4597` \
`$ git checkout pull/4597`

Update a local copy of the PR: \
`$ git checkout pull/4597` \
`$ git pull https://git.openjdk.java.net/jdk pull/4597/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4597`

View PR using the GUI difftool: \
`$ git pr show -t 4597`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4597.diff">https://git.openjdk.java.net/jdk/pull/4597.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/aarch64/atomic_aarch64.hpp [hunk 1]
```diff
@@ -46,4 +46,9 @@ extern aarch64_atomic_stub_t aarch64_atomic_cmpxchg_1_relaxed_impl;
 extern aarch64_atomic_stub_t aarch64_atomic_cmpxchg_4_relaxed_impl;
 extern aarch64_atomic_stub_t aarch64_atomic_cmpxchg_8_relaxed_impl;
 
+extern aarch64_atomic_stub_t aarch64_atomic_cmpxchg_4_seq_cst_impl;
+extern aarch64_atomic_stub_t aarch64_atomic_cmpxchg_8_seq_cst_impl;
+extern aarch64_atomic_stub_t aarch64_atomic_cmpxchg_4_release_impl;
+extern aarch64_atomic_stub_t aarch64_atomic_cmpxchg_8_release_impl;
```
Comment (shipilev) [inline src/hotspot/cpu/aarch64/atomic_aarch64.hpp:52]: Stylistic: I'd keep the ordering hierarchy here, first `release`, then `seq_cst`.

```suggestion
extern aarch64_atomic_stub_t aarch64_atomic_cmpxchg_4_release_impl;
extern aarch64_atomic_stub_t aarch64_atomic_cmpxchg_8_release_impl;
extern aarch64_atomic_stub_t aarch64_atomic_cmpxchg_4_seq_cst_impl;
extern aarch64_atomic_stub_t aarch64_atomic_cmpxchg_8_seq_cst_impl;
```

**Review by adinn — APPROVED** [review]
All looks good to me apart from the bsd changes which I am in not really in a position to comment on. I'll have to take your word for it that the builtin CAS requires the trailing mode arguments you specify.

File: src/hotspot/os_cpu/linux_aarch64/atomic_linux_aarch64.hpp [hunk 2]
```diff
@@ -151,6 +151,11 @@ inline T Atomic::PlatformCmpxchg<4>::operator()(T volatile* dest,
   switch (order) {
   case memory_order_relaxed:
     stub = aarch64_atomic_cmpxchg_4_relaxed_impl; break;
+  case memory_order_release:
+    stub = aarch64_atomic_cmpxchg_4_release_impl; break;
+  case memory_order_seq_cst:
+  case memory_order_acq_rel:
```
Comment (shipilev) [inline src/hotspot/os_cpu/linux_aarch64/atomic_linux_aarch64.hpp:157]: Swap to keep the strength hierarchy.

```suggestion
  case memory_order_acq_rel:
  case memory_order_seq_cst:
```

File: src/hotspot/os_cpu/linux_aarch64/atomic_linux_aarch64.hpp [hunk 3]
```diff
@@ -169,6 +174,11 @@ inline T Atomic::PlatformCmpxchg<8>::operator()(T volatile* dest,
   switch (order) {
   case memory_order_relaxed:
     stub = aarch64_atomic_cmpxchg_8_relaxed_impl; break;
+  case memory_order_release:
+    stub = aarch64_atomic_cmpxchg_8_release_impl; break;
+  case memory_order_seq_cst:
+  case memory_order_acq_rel:
```
Comment (shipilev) [inline src/hotspot/os_cpu/linux_aarch64/atomic_linux_aarch64.hpp:180]: Swap to keep strength hierarchy. 

```suggestion
  case memory_order_acq_rel:
  case memory_order_seq_cst:
```

File: src/hotspot/os_cpu/linux_aarch64/atomic_linux_aarch64.S [hunk 4]
```diff
@@ -112,7 +112,55 @@ aarch64_atomic_cmpxchg_8_default_impl:
...
 
-        .globl aarch64_atomic_cmpxchg_1_relaxed_default_impl
+        .globl aarch64_atomic_cmpxchg_4_release_default_impl
+        .align 5
+aarch64_atomic_cmpxchg_4_release_default_impl:
+        prfm    pstl1strm, [x0]
+0:      ldxr    w3, [x0]
+        cmp     w3, w1
+        b.ne    1f
+        stlxr    w8, w2, [x0]
```
Comment (shipilev) [inline src/hotspot/os_cpu/linux_aarch64/atomic_linux_aarch64.S:122]:
```suggestion
        stlxr   w8, w2, [x0]
```

File: src/hotspot/os_cpu/linux_aarch64/atomic_linux_aarch64.S [hunk 5]
```diff
@@ -112,7 +112,55 @@ aarch64_atomic_cmpxchg_8_default_impl:
...
+        ret
+
+        .globl aarch64_atomic_cmpxchg_8_release_default_impl
+        .align 5
+aarch64_atomic_cmpxchg_8_release_default_impl:
+        prfm    pstl1strm, [x0]
+0:      ldxr    x3, [x0]
+        cmp     x3, x1
+        b.ne    1f
+        stlxr    w8, x2, [x0]
```
Comment (shipilev) [inline src/hotspot/os_cpu/linux_aarch64/atomic_linux_aarch64.S:134]:
```suggestion
        stlxr   w8, x2, [x0]
```

File: src/hotspot/os_cpu/linux_aarch64/atomic_linux_aarch64.S [hunk 6]
```diff
@@ -112,7 +112,55 @@ aarch64_atomic_cmpxchg_8_default_impl:
...
+        cbnz    w8, 0b
+1:      mov     x0, x3
+        ret
+
+        .globl aarch64_atomic_cmpxchg_4_seq_cst_default_impl
+        .align 5
+aarch64_atomic_cmpxchg_4_seq_cst_default_impl:
+        prfm    pstl1strm, [x0]
+0:      ldaxr    w3, [x0]
```
Comment (shipilev) [inline src/hotspot/os_cpu/linux_aarch64/atomic_linux_aarch64.S:143]:
```suggestion
0:      ldaxr   w3, [x0]
```

File: src/hotspot/os_cpu/linux_aarch64/atomic_linux_aarch64.S [hunk 7]
```diff
@@ -112,7 +112,55 @@ aarch64_atomic_cmpxchg_8_default_impl:
...
+        cmp     w3, w1
+        b.ne    1f
+        stlxr    w8, w2, [x0]
```
Comment (shipilev) [inline src/hotspot/os_cpu/linux_aarch64/atomic_linux_aarch64.S:146]:
```suggestion
        stlxr   w8, w2, [x0]
```

File: src/hotspot/os_cpu/linux_aarch64/atomic_linux_aarch64.S [hunk 8]
```diff
@@ -112,7 +112,55 @@ aarch64_atomic_cmpxchg_8_default_impl:
...
+        cbnz    w8, 0b
+1:      mov     w0, w3
+        ret
+
+        .globl aarch64_atomic_cmpxchg_8_seq_cst_default_impl
+        .align 5
+aarch64_atomic_cmpxchg_8_seq_cst_default_impl:
+        prfm    pstl1strm, [x0]
+0:      ldaxr    x3, [x0]
```
Comment (shipilev) [inline src/hotspot/os_cpu/linux_aarch64/atomic_linux_aarch64.S:155]:
```suggestion
0:      ldaxr   x3, [x0]
```

File: src/hotspot/os_cpu/linux_aarch64/atomic_linux_aarch64.S [hunk 9]
```diff
@@ -112,7 +112,55 @@ aarch64_atomic_cmpxchg_8_default_impl:
...
+        cmp     x3, x1
+        b.ne    1f
+        stlxr    w8, x2, [x0]
```
Comment (shipilev) [inline src/hotspot/os_cpu/linux_aarch64/atomic_linux_aarch64.S:158]:
```suggestion
        stlxr   w8, x2, [x0]
```

**Review by shipilev — COMMENTED** [review]
Only stylistic nits. I am running the Shenandoah performance tests now.

**Comment by shipilev** [comment]
Also, if you want all platforms built without errors with GHA, you need to merge in current master.

File: src/hotspot/os_cpu/linux_aarch64/atomic_linux_aarch64.S (hunk 4 above)
Comment (theRealAph) [reply src/hotspot/os_cpu/linux_aarch64/atomic_linux_aarch64.S:122]: Well spotted!

**Review by shipilev — APPROVED** [review]
Performance data looks good, I think it does what it should.

**Review by stooart-mon — COMMENTED** [review]
I can find no fault in this - this will be a useful change.

**Pull request closed without merging**